### ISSUE TITLE
Email and Pagerduty Setup

### ISSUE THREAD

**Issue by atulkhaire** [livingsocial/rearview#25]
Hi, For setting up email alerts do I just have to configure SMTP settings or have to create a new Mailer as well ?  Sorry if its a dumb question as I am new to ruby and rails so not sure how things work.

For SMTP, in config/environments/production.rb I have added these configs.

  config.action_mailer.raise_delivery_errors = true
  config.action_mailer.delivery_method = :smtp
  config.action_mailer.smtp_settings = {
  :address              => "email-smtp.us-east-1.amazonaws.com",
  :port                    => 25,
  :user_name          => "myusername",
  :password             => "mypassword",
  :authentication       => :login,
  :enable_starttls_auto => true  }

Also could you give insights on setting up pagerduty ?

Thanks in advance

**Comment by talbright** [comment]
It depends on what you mean by "creating a mailer". In a strictly Rails sense of the word, there is already a mailer to send alerts, so no you don't need to do that. You just have to configure rails correctly (and have an smtp server available). This guide can tell you what you need to know about that: http://guides.rubyonrails.org/action_mailer_basics.html#action-mailer-configuration

I've not used Pagerduty enough to be able to tell you how to set that up. Once you have it working though just add an alert key such as the following in the UI, and it will send an alert through pager duty:

```
pagerduty:54232f6f4c4447efb6d15e20dbb7349b
```

Where _54232f6f4c4447efb6d15e20dbb7349b_ is replaced with your pager duty API key.

**Comment by atulkhaire** [comment]
Hi Trent,

Email alert is being sent but it is sent only once i.e first time when the application starts. My monitor triggers (i.e raises) alert after every 1 minute but emails are not sent after the first alert. I think since emails are being delivered that means configuration for emails is correct. 

Configured my SMTP settings as per http://guides.rubyonrails.org/action_mailer_basics.html#action-mailer-configuration.

This is something relevent I found in logs:
Rearview::ResultsHandler:0x71bc6de5 firing event :failed for : <Rearview::Job id: 4, created_at: "2014-01-29 17:28:51", updated_at: "2014-01-30 09:46:02", name: "AuthUser", active: true, last_run: "2014-01-30 09:48:01", cron_expr: "0 \* \* \* \* ?", status: "failed", user_id: 1, alert_keys: ["mailto:xyz@ab.com"], deleted_at: nil, error_timeout: 60, next_run: nil, description: "", app_id: 1, metrics: ["my-metric-name"], monitor_expr: "raise \"warning iphone_auth  hahkab\"", minutes: 1000, to_date: "now"

I am in Pacific time should that make any difference ?

Regarding Pagerduty, I entered pagerduty:my-correct-api-key but it throwed this error,
ActiveRecord::RecordInvalid (Validation failed: Alert keys pagerduty URI is invalid):

also, was not able to save monitor when I added pagerduty entry.

I have my pagerduty account at https://meadp.pagerduty.com. Do I need to change 'pagerduty_url' ?

this was the output after config verification step.

using "production" configuration:
authentication={:strategy=>:database}
default_from=xyz@ab.com
default_url_options={:host=>"localhost", :port=>"3000"}
enable_alerts=false
enable_monitor=false
enable_stats=false
graphite_url=https://www.hostedgraphite.com/avaefe1aada4/17fh854a-cgdc-424b-9F58-cA3dcf13a0w41d/graphite/render
logger=#ActiveSupport::Logger:0x5e7frw4e
pagerduty_url=https://events.pagerduty.com/generic/2010-04-15/create_event.json
preload_jobs=false
sandbox_dir=/opt/rearview-1.1.2/sandbox
sandbox_exec=["rvm-exec", "ruby-1.9.3@rearview-sandbox", "ruby"]
sandbox_timeout=10
statsd_connection=nil
verify=false
validating...PASSED

 Thanks,

**Comment by atulkhaire** [comment]
email alerts got working but pagerduty still not working.

**Comment by grimm26** [comment]
Can there be a wiki entry or something for the config you did for email?

**Comment by talbright** [comment]
Pagerduty configuration and integration is pretty minimalistic. You specify your pagerduty API key within the monitor alert keys area:

```
pagerduty:54232f6f4c4447efb6d15e20dbb7349b
```

Note that we require the API key to be a 32 character hex string.

Then you can specify the pagerduty API endpoint (config.pagerduty_url), which is defaulted to the following:

   https://events.pagerduty.com/generic/2010-04-15/create_event.json

The code that actually submits the alert is here:

http://github.com/livingsocial/rearview-engine/blob/master/lib/rearview/alerts/pagerduty_alert.rb#L10

It could be that newer accounts have to use a different endpoint, with a different payload. I'm not really sure.

**Comment by grimm26** [comment]
Hmm, I still cant get emails to send and event status is always failed.  graph seems to be working fine, though.

**Comment by grimm26** [comment]
I found that I had to set smt settings :enable_starttls_auto to false to get emails to send.  I created a new monitor, and it is sending emails just fine but a monitor that I had defined previous to this fix will still not send emails.

**Comment by talbright** [comment]
For now just delete that monitor and create a new one would be the quickest thing.

**Comment by talbright** [comment]
@swapsjd did you get pagerduty working?

**Comment by atulkhaire** [comment]
Hi Trent,

Yes Pagerduty works for me. Adding pagerduty: service-api-key in the alert section and changing pagerduty_url does the job.  Sometime soon I will add the details of SMTP settings for emails to get delivered successfully.

Does rearview support delete dashboard functionality ?

**Comment by talbright** [comment]
That'd be great if you could share your smtp settings. Might be useful for others. What did you have to change the URL to? Kinda curious why the default one didn't work.

No, there is no way to delete a dashboard. If that is something you'd like to see you can open an issue for it.

**Comment by atulkhaire** [comment]
Hi Trent,

I am using Amazon SES for email service.

These are the SMTP settings I had to add in production.rb file.

config.action_mailer.perform_deliveries = true
config.action_mailer.delivery_method = :smtp
config.action_mailer.smtp_settings = {
  :address          => "email-smtp.us-west-2.amazonaws.com",
  :port                 => 25,
  :user_name      => "username",
  :password         => "passwd",
  :authentication       => :login,
  :enable_starttls_auto => true  }

For pagerduty, I had set up my pagerduty account on "https://meadp.pagerduty.com/" 
hence setting,
config.pagerduty_url="https://meadp.pagerduty.com/" worked totally fine.

I have setup 'rearview' on centos using automation tool 'Ansible' which installs rearview and adds these configuration changes in one go. I will add an entry for Ansible scripts pretty soon. That might help some one.
